src/Stratis.Bitcoin/Consensus/ConsensusManger.cs

```diff
@@ -1,70 +1,113 @@
+
+        private readonly object treeLock;
 
-        private AsyncQueue<(ChainedHeader headerFrom, ChainedHeader headerTo)> toDownloadQueue;
+        private readonly object blockRequestedLock;
+
+        private readonly object reorgLock;
 
         public ConsensusManager(
             Network network, 
             ILoggerFactory loggerFactory, 
             IChainState chainState, 
-            IChainedHeaderValidator chainedHeaderValidator, 
+            IBlockValidator blockValidator, 
             ICheckpoints checkpoints, 
             ConsensusSettings consensusSettings, 
-            ConcurrentChain concurrentChain, 
-            ILookaheadBlockPuller lookaheadBlockPuller,
+            IBlockPuller blockPuller,
             IConsensusRules consensusRules,
-            IFinalizedBlockHeight finalizedBlockHeight)
+            IFinalizedBlockHeight finalizedBlockHeight,
+            IConnectionManager connectionManager,
+            IBlockStore blockStore = null)
         {
             this.network = network;
             this.chainState = chainState;
+            this.blockValidator = blockValidator;
             this.consensusSettings = consensusSettings;
-            this.concurrentChain = concurrentChain;
-            this.lookaheadBlockPuller = lookaheadBlockPuller;
+            this.blockPuller = blockPuller;
             this.consensusRules = consensusRules;
+            this.connectionManager = connectionManager;
+            this.blockStore = blockStore;
             this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
 
-            this.chainedHeaderTree = new ChainedHeaderTree(network, loggerFactory, chainedHeaderValidator, checkpoints, chainState, finalizedBlockHeight, consensusSettings);
+            this.chainedHeaderTree = new ChainedHeaderTree(network, loggerFactory, blockValidator, checkpoints, chainState, finalizedBlockHeight, consensusSettings);
+
+            this.treeLock = new object();
+            this.reorgLock = new object();
+            this.blockRequestedLock = new object();
 
-            this.blocksRequested = new Dictionary<uint256, List<Action>>();
-            this.toDownloadQueue = new AsyncQueue<(ChainedHeader headerFrom, ChainedHeader headerTo)>();
+            this.blocksRequested = new Dictionary<uint256, List<OnBlockDownloadedCallback>>();
+            this.toDownloadQueue = new Queue<BlockDownloadRequest>();
         }
 
-        public async Task InitializeAsync()
+        /// <summary>
+        /// Set the tip of <see cref="ConsensusManager"/> to be the header that is stored to disk,
+        /// if the given <see cref="consensusTip"/> is not in store then rewind store till a tip is found.
+        /// If <see cref="IChainState.BlockStoreTip"/> is not null (block store is enabled) ensure the store and consensus manager are on the same tip,
+        /// if consensus manager is ahead then reorg it to be the same height as store.
```

Comment:
Basically you repeated what you wrote in the first sentence. However, you are missing to describe what happens if the blocks store is not enabled.

Reply:
I will put this in a remarks then